core-js/src/definitions.ts

```diff
@@ -0,0 +1,1459 @@
+      }
+    } else {
+      this._types.set(type.name, type);
+    }
+    SchemaElement.prototype['setParent'].call(type, this);
+    return type;
+  }
+
+  /**
+   * All the directive defined on this schema _excluding_ the built-in directives, unless explicitly requested.
+   */
+  *directives(includeBuiltIns: boolean = false): Generator<DirectiveDefinition, void, undefined> {
+    if (includeBuiltIns) {
+      yield* this._builtInDirectives.values();
+    }
+    yield* this._directives.values();
+  }
+
+  /**
+   * All the built-in directives for this schema (those that are not displayed when printing the schema).
+   */
+  builtInDirectives(): IterableIterator<DirectiveDefinition> {
+    return this._builtInDirectives.values();
+  }
+
+  directive(name: string): DirectiveDefinition | undefined {
+    const directive = this._directives.get(name);
+    return directive ? directive : this._builtInDirectives.get(name);
+  }
+
+  *allSchemaElement(): Generator<SchemaElement<any, any>, void, undefined> {
+    yield this._schemaDefinition;
+    for (const type of this.types()) {
+      yield type;
+      yield* type.allChildElements();
+    }
+    for (const directive of this.directives()) {
+      yield directive;
+      yield* directive.arguments();
+    }
+  }
+
+  addDirectiveDefinition(name: string): DirectiveDefinition;
+  addDirectiveDefinition(directive: DirectiveDefinition): DirectiveDefinition; addDirectiveDefinition(directiveOrName: string | DirectiveDefinition): DirectiveDefinition {
+    const definition = typeof directiveOrName === 'string' ? new DirectiveDefinition(directiveOrName) : directiveOrName;
+    if (this.directive(definition.name)) {
+      throw buildError(`Directive ${definition} already exists in this schema`);
+    }
+    if (definition.parent) {
+      // For convenience, let's not error out on adding an already added directive.
+      if (definition.parent == this) {
+        return definition;
+      }
+      throw buildError(`Cannot add directive ${definition} to this schema; it is already attached to another schema`);
+    }
+    if (definition.isBuiltIn) {
+      if (!this.isConstructed) {
+        this._builtInDirectives.set(definition.name, definition);
+      } else {
+        throw buildError(`Cannot add built-in ${definition} to this schema (built-ins can only be added at schema construction time)`);
+      }
+    } else {
+      this._directives.set(definition.name, definition);
+    }
+    SchemaElement.prototype['setParent'].call(definition, this);
+    return definition;
+  }
+
+  clone(builtIns?: BuiltIns): Schema {
+    const cloned = new Schema(builtIns ?? this.builtIns);
+    copy(this, cloned);
+    return cloned;
+  }
+}
+
+export class SchemaDefinition extends SchemaElement<Schema, never>  {
+  readonly kind: 'SchemaDefinition' = 'SchemaDefinition';
+  protected readonly _roots: Map<SchemaRoot, ObjectType> = new Map();
+
+  get coordinate(): string {
+    return '';
```

Comment:
I think this is another sign `SchemaDefinition` should not extend `SchemaElement`.

Reply:
See comments above. But essentially, `SchemaElement` in this code is a lot more than just the coordinate, so the tradeoff is:
1. do we prefer slightly abusing the coordinate in that one case given than `SchemaDefinition` shares all other properties with other `SchemaElement`.
2. or do we insist on sticking with `SchemaElement` for things that strictly have coordinates in the graphql sense and creates a new abstraction for the other stuffs.

I've felt slightly abusing the coordinate was worth saving an abstraction. Might be the wrong choice (and I don't deny the empty coordinate thing is a tad ugly).